servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/tests/AbstractRelativeReferences.java

```diff
@@ -204,8 +205,24 @@ public void assignReferenceFromAmbiguous() {
   @Test
   @NessieApiVersions(versions = NessieApiVersion.V2)
   public void assignReferenceFromAmbiguousDetached() {
-    NessieError error = expectError(rest().body(etl1).put("trees/@~1"), 400);
-    checkError(error, BAD_REQUEST, "Expected hash must contain a starting commit ID.");
+    NessieError error =
+        expectError(
+            rest()
+                .body(Detached.of("~1"))
+                .put("trees/{ref}@{hash}", etl1.getName(), etl1.getHash()),
+            400);
+    checkError(error, BAD_REQUEST, "Target hash must contain a starting commit ID.");
+  }
+
+  /**
+   * The {@code @~1} ref spec means a "detached" reference, which is not a named reference, hence it
+   * cannot be assigned.
+   */
+  @Test
+  @NessieApiVersions(versions = NessieApiVersion.V2)
+  public void assignDetachedReferenceDeclined() {
+    NessieError error = expectError(rest().body(etl1).put("trees/@~1"), 404);
+    checkError(error, REFERENCE_NOT_FOUND, "Named reference 'DETACHED' not found");
```

Comment:
Let's approach this from a different angle. A delete reference call with the same URL path currently produces `Named reference 'DETACHED' not found`.

I think both assign and delete reference end points should use the same reference lookup logic.

That said, `DETACHED` as a name cannot be use to create a reference. Even though `DetachedRef extends NamedRef`, as the API level `ReferenceType` only has `BRANCH` and `TAG`. This is asymmetrical, but essentially at the API level `DETACHED` references are not "named references". For example, they cannot be obtained from `getReferenceByName`.

Reply:
Also, a request to `put("trees/@1122334455667788")` without this PR leads to `Argument must be a branch or tag, but was DetachedRef{}`. I think a `Named reference 'DETACHED' not found` error message is clearer.

Reply:
Let's sort this out in a follow-up PR.

Reply:
OK, I am going to approve since the main intent of the PR is valid. 

But indeed I would like us to come up, if possible, with a better solution for resolving detached refs and presenting errors to the client. The whole point of #7268 was to give a consistent behavior to all endpoints regarding resolution of named refs and relative hashes. I find it a bit disappointing to special-case one or two endpoints again.

I think we should instead revisit `resolveHashOnRef`. Maybe there are other endpoints where we should disallow detached refs completely? For example, I suspect that commit, merge and transplant endpoints could benefit from the same common handling of non-branch refs: for now, they all have special checks that throw HTTP 400 with message "Can only commit into branches."

The ideal error message in such cases would imho be HTTP 400 and a message like `Reference (to assign|to delete|to commit into|etc.) must be a branch, but was: (detached|tag)` or stg in the style.

With this refactoring, the `checkArgument` call could certainly be removed, since it does more or less the same of what `HashValidator.EXPECTED_HASH` is already supposed to do.

I can look into this follow-up work.